samples/guide/src/main/java/okhttp3/recipes/kt/WiresharkExample.kt

```diff
@@ -0,0 +1,246 @@
+
+      if (masterSecretHex != null) {
+        val keyLog = "CLIENT_RANDOM $random $masterSecretHex"
+
+        println(keyLog)
+        File("/tmp/key.log").appendText("$keyLog\n")
+      }
+    }
+
+    random = null
+  }
+
+  class Factory : EventListener.Factory {
+    override fun create(call: Call): EventListener {
+      return WireSharkKeyLoggerListener()
+    }
+  }
+
+  companion object {
+    private lateinit var logger: Logger
+
+    private val SSLSession.masterSecret: SecretKey?
+      get() = javaClass.getDeclaredField("masterSecret")
+          .apply {
+            isAccessible = true
+          }
+          .get(this) as? SecretKey
+
+    val randomRegex = "\"random\"\\s+:\\s+\"([^\"]+)\"".toRegex()
+
+    fun register() {
+      // Enable JUL logging for SSL events.
+      System.setProperty("javax.net.debug", "")
+      logger = Logger.getLogger("javax.net.ssl")
+          .apply {
+            level = Level.FINEST
+          }
+    }
+  }
+}
+
+class WiresharkExample {
+  val connectionSpec =
+    Builder(ConnectionSpec.RESTRICTED_TLS)
+        .tlsVersions(TLS_1_2)
+//        .cipherSuites(CipherSuite.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256)
+        .build()
+
+  var bootstrapClient = OkHttpClient.Builder()
+      .connectionSpecs(listOf(connectionSpec))
+      .addInterceptor(BrotliInterceptor)
+      .eventListenerFactory(WireSharkKeyLoggerListener.Factory())
+      .build()
+
+  val dns = DnsOverHttps.Builder().client(bootstrapClient)
+    .url("https://1.1.1.1/dns-query".toHttpUrl())
+    .includeIPv6(false)
+    .build()
+
+  val client = OkHttpClient.Builder()
+      .connectionSpecs(listOf(connectionSpec))
+      .eventListenerFactory(WireSharkKeyLoggerListener.Factory())
+      .addInterceptor(BrotliInterceptor)
+      .dns(dns)
+      .build()
+
+  fun run() {
+    val fbRequest = Request.Builder()
+        .url("https://graph.facebook.com/robots.txt?s=fb")
+        .build()
+    val twitterRequest = Request.Builder()
+        .url("https://api.twitter.com/robots.txt?s=tw")
+        .build()
+    val googleRequest = Request.Builder()
+        .url("https://www.google.com/robots.txt?s=g")
+        .build()
+
+    for (i in 1..2) {
+      sendTestRequest(fbRequest)
+      sendTestRequest(twitterRequest)
+      sendTestRequest(googleRequest)
+
+      Thread.sleep(2000)
+    }
+  }
+
+  private fun sendTestRequest(twitterRequest: Request) {
+    client.newCall(twitterRequest)
+        .execute()
+        .use {
+          val firstLine = it.body!!.string()
+              .lines()
+              .first()
+          println("${it.code} ${it.request.url.host} $firstLine")
+          Unit
+        }
+  }
+}
+
+fun main() {
+  println("TLSv1.2 traffic will be logged automatically")
+  println("Follow instructions at https://github.com/neykov/extract-tls-secrets for TLSv1.3")
+  println("Pid: ${ProcessHandle.current().pid()}")
+
+  Thread.sleep(10000)
+
+  WireSharkKeyLoggerListener.register()
```

Comment:
Should the event listener do this automatically? I want to paste the whole thing somewhere, right?

Reply:
It needs to be done ASAP, ideally on the command line via -Djavax.net.debug=